### ISSUE TITLE
Loading plugin before deserializing the TensorRT model as it uses the plugin library.

### ISSUE THREAD

**Issue by TecntOvO** [ultralytics/ultralytics#21674]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Refer to the follwing link:
https://forums.developer.nvidia.com/t/python-tensorrt-loaded-engine-failed/250981
, which is the problem encountered in my project when loading my Tensort model.

### Environment

Ultralytics 8.3.170  Python-3.8.20 torch-2.1.0a0+41361538.nv23.06 CUDA:0 (Xavier, 14885MiB)
Setup complete ✅ (6 CPUs, 14.5 GB RAM, 41.9/102.1 GB disk)

OS Linux-5.10.216-tegra-aarch64-with-glibc2.26
Environment Linux
Python 3.8.20
Install pip
RAM 14.54 GB
Disk 41.9/102.1 GB
CPU ARMv8 Processor rev 0 (v8l)
CPU count 6
GPU Xavier, 14885MiB
GPU count 1
CUDA 11.4

### Minimal Reproducible Example
The link has provided a clear example of how to reproduce the error. The following example is how the bug is triggled in my project

```
from ultralytics import YOLO

if __name__ == '__main__':
    model = YOLO(r"improve_v11.engine")
    model.predict(source=r"D:\YOLOv11\dataset\split_final\images\val",save=True, line_width=2)
```

The file improve_v11.engine was exported by yolo.export from my pytorch weigh file in Jetson Xavier
 Nano. As I run the code in my Jetson, when the system start to load model in autobackend.py in ultralytics/nn, the function 

`model = runtime.deserialize_cuda_engine(f.read())  # read engine`

throws an error as follows:

```
 [TRT] [E] 1: [pluginV2Runner.cpp::load::292] Error Code 1: Serialization (Serialization assertion creator failed.Cannot deserialize plugin since corresponding IPluginCreator not found in Plugin Registry)
 [TRT] [E] 4: [runtime.cpp::deserializeCudaEngine::49] Error Code 4: Internal Error (Engine deserialization failed.)
```

When I mimic the solution in the link and add the function to load plugin, the problem is solved.

```
# This part of the code is located on line 348 of the autobackend.py in ultralytics/nn
logger = trt.Logger(trt.Logger.INFO)
**trt.init_libnvinfer_plugins(logger, namespace="") # init the plugin before deserializing the model as it uses the plugin library.**
# Read file
with open(w, "rb") as f, trt.Runtime(logger) as runtime:
    try:
        meta_len = int.from_bytes(f.read(4), byteorder="little")  # read metadata length
        metadata = json.loads(f.read(meta_len).decode("utf-8"))  # read metadata
        dla = metadata.get("dla", None)
        if dla is not None:
            runtime.DLA_core = int(dla)
    except UnicodeDecodeError:
        f.seek(0)  # engine file may lack embedded Ultralytics metadata
    model = runtime.deserialize_cuda_engine(f.read())  # read engine
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TecntOvO, thank you for your interest in Ultralytics 🚀! This is an automated response to help get things moving—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the Docs for new users where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

Since this appears to be a 🐛 Bug Report related to TensorRT plugin initialization during Predict, please provide a minimum reproducible example (MRE) to help us debug it:
- MRE Guide: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our Tips for Best Training Results:
- Training Tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat: https://discord.com/invite/ultralytics 🎧
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit for threads and knowledge sharing: https://www.reddit.com/r/Ultralytics/

## Upgrade
Upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us dig in quickly, please share:
- A concise MRE script showing exactly how you load and run the TensorRT engine with Ultralytics Predict
- The exact engine build details (TensorRT version, opset, dynamic/explicit batch, plugins used)
- Confirmation of any plugin library loading steps and their order relative to runtime.deserialize_cuda_engine
- Full error traceback and logs
- Whether the issue reproduces after pip install -U ultralytics in a clean environment

Thanks for the detailed environment info and snippet you’ve already included 🙌

**Comment by Y-T-G** [comment]
Can you provide the minimum reproducible example? To generate the engine and reproduce the error.

**Comment by TecntOvO** [comment]
> Can you provide the minimum reproducible example? To generate the engine and reproduce the error.

Of course, I have complemented the Minimal Reproducible Example

**Comment by Y-T-G** [comment]
Is this the YOLO11 model? Or have you modified the architecture?

**Comment by TecntOvO** [comment]
> Is this the YOLO11 model? Or have you modified the architecture?

The TensorRT model is transformed from my improved YOLOv11 model which is added some spatial mechanisms. When I transform my improved model into onnx format, there is nothing wrong, so is the process of tranforming the onnx file into TensorRT format. However, the error does not occur as I run the prediction on the TensorRT file transformed from original YOLOv11 model. By adding the provided function, both situations run successfully.

**Comment by Y-T-G** [comment]
I think it's caused by your modifications since the default models don't have an issue
